src/Pact/Eval.hs

```diff
@@ -436,6 +467,82 @@ resolveGovernance solvedDefs m' = fmap MDModule $ forM m' $ \g -> case g of
       Nothing -> evalError (_tInfo g) "Unknown module governance reference"
     _ -> evalError (_tInfo g) "Invalid module governance, should be var"
 
+-- Auxiliary data types for strict fold on heap cost.
+data HeapMemState
+  = HeapMemState
+  { _hmMemoEnv :: !(M.Map Name Bytes)
+  , _hmTotalMem :: !Bytes
+  }
+
+data HeapFold
+  = HeapFold
+  { _hfAllDefs :: !(HM.HashMap Text Ref)
+  , _hfMemoEnv :: !(M.Map Name Bytes)
+  , _hfTotalMem :: !Bytes
+  , _hfLastGas :: !Gas
+  }
+
+
+-- Inline the defuns according to the set heap limit for modules.
+-- We keep a memoized cost of each inlined `defun` so as to not calculate `sizeOf` more than once
+-- per inlined function.
+dresolveMem
+  :: Gas
+  -> Info
+  -> HeapFold
+  -> (Term (Either Text (Ref' (Term Name))), Text, c)
+  -> Eval e HeapFold
+dresolveMem gasLimit info (HeapFold allDefs costMemoEnv currMem _) (defTerm, defName, _) = do
+  let
+    (!unified, (HeapMemState costMemoEnv' totalMem))
+      = flip runState (HeapMemState costMemoEnv (sizeOf defTerm+currMem)) $ traverse (unifyMemo allDefs) defTerm
+  case unified of
+    t@TConst{} -> do
+      t' <- runSysOnly $ evalConstsNonRec (Ref t)
+      lastGas <- checkExtraGas totalMem
+      pure (HeapFold (HM.insert defName t' allDefs) costMemoEnv' totalMem lastGas)
+    _ -> do
+      lastGas <- checkExtraGas totalMem
+      pure (HeapFold (HM.insert defName (Ref unified) allDefs) costMemoEnv' totalMem lastGas)
+  where
+  checkExtraGas tm = do
+    GasEnv {..} <- view eeGasEnv
+    !g0 <- use evalGas
+    let !gUsed = g0 + runGasModel _geGasModel "moduleMem" (GModuleMemory tm)
+    if gUsed > gasLimit then
+      throwErr GasError info $ "Gas limit (" <> pretty gasLimit <> ") exceeded: " <> pretty gUsed
+    else return gUsed
+  -- Inline a foreign defun: memoize the cost, since it may be expensive to calculate
+  unifyMemo _ (Right (Ref td@(TDef defn _))) = do
+    let (DefName defname) = _dDefName defn
+        name = QName (QualifiedName (_dModule defn) defname def)
+    memoEnv <- gets _hmMemoEnv
+    case M.lookup name memoEnv of
+      Just heapCost -> do
+        modify' (\(HeapMemState env total) -> HeapMemState env (total + heapCost))
+        pure (Ref td)
+      Nothing -> do
+        let !heapCost = sizeOf td
+        modify' (\(HeapMemState env total) -> HeapMemState (M.insert name heapCost env) (total+heapCost))
+        pure (Ref td)
+  -- Note: inlining only ever inlines tdefs and modrefs, it's fine to not charge
```

Comment:
Also `TConsts`, `TTable` and `TSchema`, missing from the comment, but these are small and non-recursive. Is it fine to keep them free?